From c911dc221f28f16bcc5cbac5863cbd9e7e391db9 Mon Sep 17 00:00:00 2001
From: CL Wang <cl634@andestech.com>
Date: Thu, 7 Sep 2023 10:34:35 +0800
Subject: [PATCH 139/191] input: andes: touchscreen: Refine the driver

1. Refine the driver to fit the coding style rule
2. Remove REG32 and use ioread32 instead.
3. Fix the wrong data type of regs from resource_size_t to "void __iomem *"
PS: This driver just passed the compilation but has not been tested yet.

Signed-off-by: CL Wang <cl634@andestech.com>
---
 drivers/input/touchscreen/Kconfig        |  17 +++
 drivers/input/touchscreen/Makefile       |   1 +
 drivers/input/touchscreen/cpe_ts_ae350.c | 150 ++++++++++++-----------
 drivers/input/touchscreen/cpe_ts_ae350.h |  11 +-
 4 files changed, 106 insertions(+), 73 deletions(-)

diff --git a/drivers/input/touchscreen/Kconfig b/drivers/input/touchscreen/Kconfig
index dc90a3ea51ee..f3cac22338c0 100644
--- a/drivers/input/touchscreen/Kconfig
+++ b/drivers/input/touchscreen/Kconfig
@@ -89,6 +89,23 @@ config TOUCHSCREEN_AD7879_SPI
 	  To compile this driver as a module, choose M here: the
 	  module will be called ad7879-spi.
 
+config TOUCHSCREEN_CPE_TS_AE350
+	tristate "Touchscreen Driver for AE350/XC7"
+	help
+	  This driver directly accesses SPI controller to communicate with
+	  the ads7846 chip.
+
+	  Once we have a SPI controller driver, we can adopt to the SPI
+	  framework that kernel provides.
+
+config TOUCHSCREEN_CPE_TS_DEJITTER_AE350
+	bool "Dejitter Detection"
+	depends on TOUCHSCREEN_CPE_TS_AE350
+	help
+	  Say Y here to enable dejitter detection in AE350/Big Orca Touchscreen Driver.
+
+	  If unsure, say y.
+
 config TOUCHSCREEN_ADC
 	tristate "Generic ADC based resistive touchscreen"
 	depends on IIO
diff --git a/drivers/input/touchscreen/Makefile b/drivers/input/touchscreen/Makefile
index 557f84fd2075..2686b68a4a32 100644
--- a/drivers/input/touchscreen/Makefile
+++ b/drivers/input/touchscreen/Makefile
@@ -116,3 +116,4 @@ obj-$(CONFIG_TOUCHSCREEN_ROHM_BU21023)	+= rohm_bu21023.o
 obj-$(CONFIG_TOUCHSCREEN_RASPBERRYPI_FW)	+= raspberrypi-ts.o
 obj-$(CONFIG_TOUCHSCREEN_IQS5XX)	+= iqs5xx.o
 obj-$(CONFIG_TOUCHSCREEN_ZINITIX)	+= zinitix.o
+obj-$(CONFIG_TOUCHSCREEN_CPE_TS_AE350)	+= cpe_ts_ae350.o
diff --git a/drivers/input/touchscreen/cpe_ts_ae350.c b/drivers/input/touchscreen/cpe_ts_ae350.c
index 565827a8e14d..1812c5753991 100755
--- a/drivers/input/touchscreen/cpe_ts_ae350.c
+++ b/drivers/input/touchscreen/cpe_ts_ae350.c
@@ -1,36 +1,42 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (C) 2023 Andes technology Corporation
+ */
+
 #include <linux/input.h>
 #include <linux/interrupt.h>
 #include <linux/platform_device.h>
 #include <linux/io.h>
-#include <asm/irq.h>
-#include "cpe_ts_ae350.h"
 #include <linux/module.h>
 #include <linux/irqnr.h>
 #include <linux/irq.h>
 #include <linux/delay.h>
 #include <linux/of.h>
 #include <linux/of_device.h>
+#include <asm/irq.h>
+#include "cpe_ts_ae350.h"
 
-#define REG32(a)        (*(volatile unsigned int *)(a))
-#define ads_dbg( enabled, tagged, ...)					\
-	do{								\
-		if( enabled){						\
-			if( tagged)					\
-				printk( "[ %30s() ] ", __func__);       \
-			printk( __VA_ARGS__);				\
+#define ads_dbg(enabled, tagged, ...)					\
+	do {								\
+		if (enabled) {						\
+			if (tagged)					\
+				pr_err("[ %30s() ] ", __func__);        \
+			pr_err(__VA_ARGS__);				\
 		}							\
-	} while( 0)
+	} while (0)
+
+#define TS_POLL_DELAY        (1 * 1000000)          /* ns delay before the first sample */
+#define TS_POLL_PERIOD       (delay * 1000000)      /* ns delay between samples */
+#define	read_reg(addr)       ioread32(addr)
+#define	write_reg(val, addr) iowrite32(val, addr)
 
-#define TS_POLL_DELAY		( 1 * 1000000)		/* ns delay before the first sample */
-#define TS_POLL_PERIOD		( delay * 1000000)	/* ns delay between samples */
-static int debug = 0;
+static int debug;
 static int delay = 25;
 module_param(debug, int, 0);
 module_param(delay, int, 0);
 
-struct ads7846
-{
-	void __iomem * regs;
+struct ads7846 {
+	void __iomem *regs;
 	struct input_dev *input;
 	char phys[32];
 	struct hrtimer timer;
@@ -39,8 +45,7 @@ struct ads7846
 	bool disabled;
 };
 
-struct ts_event
-{
+struct ts_event {
 	int x;
 	int y;
 	int z1, z2;
@@ -49,37 +54,42 @@ struct ts_event
 
 struct ads7846 touchscreen;
 
-#define ADS_START		( 0x1UL << 7)
-#define ADS_A2A1A0_d_y		( 0x1UL << 4)	/* differential */
-#define ADS_A2A1A0_d_z1		( 0x3UL << 4)	/* differential */
-#define ADS_A2A1A0_d_z2		( 0x4UL << 4)	/* differential */
-#define ADS_A2A1A0_d_x		( 0x5UL << 4)	/* differential */
-#define ADS_12_BIT		( 0x0UL << 3)
-#define ADS_SER			( 0x1UL << 2)	/* non-differential */
-#define ADS_DFR			( 0x0UL << 2)	/* differential */
-#define ADS_PD10_PDOWN		( 0x0UL << 0)	/* lowpower mode + penirq */
-#define ADS_PD10_ADC_ON		( 0x1UL << 0)	/* ADC on */
-#define ADS_PD10_REF_ON		( 0x2UL << 0)	/* vREF on + penirq */
-#define ADS_PD10_ALL_ON		( 0x3UL << 0)	/* ADC + vREF on */
-#define MAX_12BIT		( ( 0x1UL << 12) - 1)
-
-#define READ_X			( ADS_A2A1A0_d_x  | ADS_12_BIT | ADS_DFR)
-#define READ_Y			( ADS_A2A1A0_d_y  | ADS_12_BIT | ADS_DFR)
-#define READ_Z1			( ADS_A2A1A0_d_z1 | ADS_12_BIT | ADS_DFR)
-#define READ_Z2			( ADS_A2A1A0_d_z2 | ADS_12_BIT | ADS_DFR)
+#define ADS_START           (0x1UL << 7)
+#define ADS_A2A1A0_d_y      (0x1UL << 4)	/* differential */
+#define ADS_A2A1A0_d_z1     (0x3UL << 4)	/* differential */
+#define ADS_A2A1A0_d_z2     (0x4UL << 4)	/* differential */
+#define ADS_A2A1A0_d_x      (0x5UL << 4)	/* differential */
+#define ADS_12_BIT          (0x0UL << 3)
+#define ADS_SER             (0x1UL << 2)	/* non-differential */
+#define ADS_DFR             (0x0UL << 2)	/* differential */
+#define ADS_PD10_PDOWN      (0x0UL << 0)	/* lowpower mode + penirq */
+#define ADS_PD10_ADC_ON     (0x1UL << 0)	/* ADC on */
+#define ADS_PD10_REF_ON     (0x2UL << 0)	/* vREF on + penirq */
+#define ADS_PD10_ALL_ON     (0x3UL << 0)	/* ADC + vREF on */
+#define MAX_12BIT           ((0x1UL << 12) - 1)
+
+#define READ_X              (ADS_A2A1A0_d_x  | ADS_12_BIT | ADS_DFR)
+#define READ_Y              (ADS_A2A1A0_d_y  | ADS_12_BIT | ADS_DFR)
+#define READ_Z1             (ADS_A2A1A0_d_z1 | ADS_12_BIT | ADS_DFR)
+#define READ_Z2             (ADS_A2A1A0_d_z2 | ADS_12_BIT | ADS_DFR)
 
 static int
 read_val(struct ads7846 *ts, unsigned long cmd)
 {
 	unsigned long data = 0;
+
 	ads_dbg(0, 1, "Queuing data: 0x%08lx\n", cmd << 16);
 
-	while (!(REG32(ts->regs + STATUS) & TX_EMPTY));
-	REG32(ts->regs + DATA) = (ADS_START | cmd) << 16;
-	REG32(ts->regs + CMD) = 1;
+	while (!(read_reg(ts->regs + STATUS) & TX_EMPTY))
+		;
+
+	write_reg((ADS_START | cmd) << 16, ts->regs + DATA);
+	write_reg(1, ts->regs + CMD);
 
-	while ((REG32(ts->regs + STATUS) & RX_EMPTY));
-	data = (REG32(ts->regs + DATA) >> 3) & 0xFFF;
+	while ((read_reg(ts->regs + STATUS) & RX_EMPTY))
+		;
+
+	data = (read_reg(ts->regs + DATA) >> 3) & 0xFFF;
 	ads_dbg(0, 1, "CMD <%02lx> data: 0x%08lx( %ld)\n", cmd, data, data);
 
 	return data;
@@ -104,24 +114,22 @@ static enum hrtimer_restart ads7846_timer(struct hrtimer *handle)
 {
 	struct ads7846 *ts = container_of(handle, struct ads7846, timer);
 	struct ts_event e;
-	static int xp = 0, yp = 0;
+	static int xp, yp;
 
 	if (ts->disabled)
 		return HRTIMER_NORESTART;
 
-	if (!pendown(ts))
-	{
+	if (!pendown(ts)) {
 		ads_dbg(debug, 1, "Release\n");
 		input_report_key(ts->input, BTN_TOUCH, 0);
 		input_report_abs(ts->input, ABS_PRESSURE, 0);
 		input_sync(ts->input);
 		enable_irq(ts->irq);
-                return HRTIMER_NORESTART;
+		return HRTIMER_NORESTART;
 	}
 	report(ts, &e);
 #ifdef CONFIG_TOUCHSCREEN_CPE_TS_DEJITTER_AE350
-	if (abs(xp - e.x) > FILTER_LIMIT || abs(yp - e.y) > FILTER_LIMIT)
-	{
+	if (abs(xp - e.x) > FILTER_LIMIT || abs(yp - e.y) > FILTER_LIMIT) {
 #endif
 		input_report_key(ts->input, BTN_TOUCH, 1);
 		input_report_abs(ts->input, ABS_X, e.x);
@@ -157,17 +165,24 @@ static irqreturn_t ads7846_irq(int irq, void *handle)
 
 static int xspi_init_hw(struct at_ts_platdata *pdata)
 {
-	int id = REG32(pdata->regs + ID_REV);
+	int id = read_reg(pdata->regs + ID_REV);
+	unsigned int reg_val;
 
-	if((id&ID_MSK) !=(ID_ATCSPI200<<ID_OFF))
-	{
+	if ((id & ID_MSK) != (ID_ATCSPI200 << ID_OFF)) {
 		ads_dbg(1, 0, "ADS7846 Touchscreen controller initialized failed:\n"
 			"\tcannot detect ANdes SPI Controller\n");
 		return -ENXIO;
 	}
-	REG32(pdata->regs + TIMING) |= (pdata->clk/(pdata->sclk<<1))-1;
-	REG32(pdata->regs + FORMAT) &= DATA_LEN_MSK;
-	REG32(pdata->regs + FORMAT) |= (23<<DATA_LEN_OFF);
+
+	reg_val = read_reg(pdata->regs + TIMING);
+	write_reg(reg_val | ((pdata->clk / (pdata->sclk << 1)) - 1),
+		  pdata->regs + TIMING);
+
+	reg_val = read_reg(pdata->regs + FORMAT);
+	write_reg(reg_val & DATA_LEN_MSK, (pdata->regs + FORMAT));
+
+	reg_val = read_reg(pdata->regs + FORMAT);
+	write_reg(reg_val | (23 << DATA_LEN_OFF), (pdata->regs + FORMAT));
 
 	return 0;
 }
@@ -184,7 +199,7 @@ static int __exit ads7846_remove(struct platform_device *pdev)
 }
 
 #ifdef CONFIG_PM
-static int ads7846_suspend( struct platform_device *pdev, pm_message_t message)
+static int ads7846_suspend(struct platform_device *pdev, pm_message_t message)
 {
 	struct ads7846 *ts = platform_get_drvdata(pdev);
 
@@ -196,7 +211,7 @@ static int ads7846_suspend( struct platform_device *pdev, pm_message_t message)
 	return 0;
 }
 
-static int ads7846_resume( struct platform_device *pdev)
+static int ads7846_resume(struct platform_device *pdev)
 {
 	struct ads7846 *ts = platform_get_drvdata(pdev);
 
@@ -207,9 +222,6 @@ static int ads7846_resume( struct platform_device *pdev)
 
 	return 0;
 }
-#else
-#define ads7846_suspend NULL
-#define ads7846_resume NULL
 #endif
 
 static const struct of_device_id ads7846_dt_ids[] = {
@@ -218,7 +230,7 @@ static const struct of_device_id ads7846_dt_ids[] = {
 	},
 	{},
 };
-MODULE_DEVICE_TABLE(of, ads7846_dt_ids);
+MODULE_DEVICE_TABLE(of, c_dt_ids);
 
 #ifdef CONFIG_OF
 static struct at_ts_platdata *ts_parse_dt(struct device *dev)
@@ -230,21 +242,18 @@ static struct at_ts_platdata *ts_parse_dt(struct device *dev)
 		return ERR_PTR(-ENOENT);
 
 	pdata = devm_kzalloc(dev, sizeof(*pdata), GFP_KERNEL);
-	if (!pdata) {
-		dev_err(dev, "failed to allocate platform data\n");
+	if (!pdata)
 		return ERR_PTR(-ENOMEM);
-	}
 
-	if (of_property_read_u32(np, "spi-max-frequency", &pdata->sclk)){
+	if (of_property_read_u32(np, "spi-max-frequency", &pdata->sclk)) {
 		dev_err(dev, "failed to get spi-max-frequency property\n");
 		return ERR_PTR(-EINVAL);
 	}
 
-	if (of_property_read_u32(np, "clock-frequency", &pdata->clk)){
+	if (of_property_read_u32(np, "clock-frequency", &pdata->clk)) {
 		dev_err(dev, "failed to get clock-frequency property\n");
 		return ERR_PTR(-EINVAL);
 	}
-	printk("pdata->sclk %d , pdata->clk %d\n",pdata->sclk,pdata->clk);
 
 	if (of_property_read_u32(np, "x-size", &pdata->x_max)) {
 		dev_err(dev, "failed to get x-size property\n");
@@ -278,15 +287,14 @@ static int ads7846_probe(struct platform_device *pdev)
 	platform_set_drvdata(pdev, ts);
 	r = platform_get_resource(pdev, IORESOURCE_MEM, 0);
 	ts->regs = devm_ioremap_resource(&pdev->dev, r);
-	pdata->regs = (resource_size_t)ts->regs;
+	pdata->regs = ts->regs;
 
 	err = xspi_init_hw(pdata);
 	if (err)
 		goto err_unmap;
 
 	input_dev = input_allocate_device();
-	if (!input_dev)
-	{
+	if (!input_dev) {
 		err = -ENOMEM;
 		goto err_free_mem;
 	}
@@ -323,16 +331,18 @@ static int ads7846_probe(struct platform_device *pdev)
 	return err;
 }
 
-static struct platform_driver ads7846_driver =
-{
+static struct platform_driver ads7846_driver = {
 	.driver = {
 		.name = "ads7846",
 		.owner	= THIS_MODULE,
 		.of_match_table = of_match_ptr(ads7846_dt_ids),
 	},
 	.remove = __exit_p(ads7846_remove),
+
+#ifdef CONFIG_PM
 	.suspend = ads7846_suspend,
 	.resume = ads7846_resume,
+#endif
 };
 
 module_platform_driver_probe(ads7846_driver, ads7846_probe);
diff --git a/drivers/input/touchscreen/cpe_ts_ae350.h b/drivers/input/touchscreen/cpe_ts_ae350.h
index 5ca9910a0968..68e0fa407782 100755
--- a/drivers/input/touchscreen/cpe_ts_ae350.h
+++ b/drivers/input/touchscreen/cpe_ts_ae350.h
@@ -1,10 +1,15 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * Copyright (C) 2023 Andes technology Corporation
+ */
+
 #ifndef SPI_ANDES_H
 #define SPI_ANDES_H
 
 #define XILINX_SPI_NAME "andes-spi"
 
 struct at_ts_platdata {
-	resource_size_t regs;
+	void __iomem *regs;
 	unsigned int clk;
 	unsigned int sclk;
 	unsigned int x_max;
@@ -24,7 +29,7 @@ struct at_ts_platdata {
 #define DATA_LEN_MSK	(0x1f<<8)
 
 /* SPI Transfer Control Register */
-#define TCONTROL		0x20
+#define TCONTROL	0x20
 #define TRANS_MODE_OFF	24
 #define TRANS_MODE_MSK	(0xf<<TRANS_MODE_OFF)
 #define WRITE_THEN_READ	(0x3<<TRANS_MODE_OFF)
@@ -67,7 +72,7 @@ struct at_ts_platdata {
 #define XSPI_CR_MASTER_MODE	0x04
 #define XSPI_CR_CPOL		0x08
 #define XSPI_CR_CPHA		0x10
-#define XSPI_CR_MODE_MASK	( XSPI_CR_CPHA | XSPI_CR_CPOL)
+#define XSPI_CR_MODE_MASK	(XSPI_CR_CPHA | XSPI_CR_CPOL)
 #define XSPI_CR_TXFIFO_RESET	0x20
 #define XSPI_CR_RXFIFO_RESET	0x40
 #define XSPI_CR_MANUAL_SSELECT	0x80
-- 
2.34.1

